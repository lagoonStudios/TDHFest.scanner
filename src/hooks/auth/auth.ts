import React from "react";
import {
  getAuth,
  onAuthStateChanged,
  User,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";

const auth = getAuth();

export function useAuthentication() {
  const [user, setUser] = React.useState<User>();

  React.useEffect(() => {
    const unsubscribeFromAuthStatuChanged = onAuthStateChanged(auth, (user) => {
      if (user) {
        // User is signed in, see docs for a list of available properties
        // https://firebase.google.com/docs/reference/js/firebase.User
        setUser(user);
      } else {
        // User is signed out
        setUser(undefined);
      }
    });

    return unsubscribeFromAuthStatuChanged;
  }, []);

  return {
    user,
  };
}

export function logIn(email: string, password: string) {
  return signInWithEmailAndPassword(auth, email, password);
}

export function logOut() {
  return signOut(auth);
}

export function handleAuthErrorMessage(message: string): string {
  console.log("error msg: ", message);

  if (message.includes("invalid-email")) return "Email inv치lido";

  else if (message.includes("invalid-login-credentials")) return "Usuario o contrase침a inv치lidos";

  return "Error de autenticaci칩n";
}
