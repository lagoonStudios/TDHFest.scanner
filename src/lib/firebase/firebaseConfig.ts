import { REACT_APP_FIREBASE_CONFIG } from "@env";
import { initializeApp } from "firebase/app";
import { getFirestore } from "firebase/firestore";

export const firebaseConfig = JSON.parse(REACT_APP_FIREBASE_CONFIG);

const firebase = initializeApp(firebaseConfig);
const firestore = getFirestore(firebase);

export { firestore, firebase };
