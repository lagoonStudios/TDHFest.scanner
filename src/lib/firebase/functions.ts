export enum Collections {
  Users = "Users",
  Tickets = "Tickets",
  TicketTypes = "TicketTypes",
  Events_PaymentTypes = "Events_PaymentTypes",
}

export const eventId = "id6jrKAOPHWG0RBkTAfa";

export const handleFirebaseErrorMessage = (errorCode: string) => {
  const errorMessageMap: { [key: string]: string } = {
    "not-found": "No encontrado",
  };
  const defaultMessage = "Algo saliÃ³ mal";

  const mappedMessage = Object.keys(errorMessageMap).find((key) => errorCode.includes(key));

  return mappedMessage ? errorMessageMap[mappedMessage] : defaultMessage;
};
