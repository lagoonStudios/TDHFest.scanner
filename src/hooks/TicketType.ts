import { firestore } from "@/lib/firebase/firebaseConfig";
import { Collections, eventId } from "@/lib/firebase/functions";
import { TicketType } from "@/models/app.models";
import { collection, getDocs, query, where } from "firebase/firestore";
import { useEffect, useState } from "react";

export function useTicketType(): TicketType[] {
  const [ticketTypes, setTicketTypes] = useState<TicketType[]>([]);
  const q = query(collection(firestore, Collections.TicketTypes), where("eventId", "==", eventId));

  useEffect(() => {
    getDocs(q).then((docsRef) => {
      const docs = docsRef.docs.map((doc) => ({ ...(doc.data() as TicketType), id: doc.id }));
      setTicketTypes(docs);
    });
  }, []);

  return ticketTypes;
}
