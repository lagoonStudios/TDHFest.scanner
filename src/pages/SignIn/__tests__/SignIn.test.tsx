import * as firebaseAuth from "firebase/auth";
import Toast from "react-native-toast-message";
import { screen } from "@testing-library/react-native";

import * as localAuth from "@/hooks/auth";
import setupComponent from "@/utils/tests-utils";

import { SignIn } from "..";

describe("<SignIn />", () => {
  test("Should render propperly / has 2 childs", () => {
    const tree = setupComponent(<SignIn />).toJSON();
    expect(tree.children.length).toBe(2);
  });

  test("Should match snapshot", () => {
    const tree = setupComponent(<SignIn />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Should mark all fields as required", async () => {
    const { user } = setupComponent(<SignIn />);

    await user.press(screen.getByTestId("submit-login"));
    const requiredTexts = screen.getAllByText("Requerido");

    expect(requiredTexts.length).toBe(2);
  });

  test("Should mark email field as invalid when typed an invalid email", async () => {
    const { user } = setupComponent(<SignIn />);

    await user.type(screen.getByTestId("email-input"), "invalidEmail");
    await user.press(screen.getByTestId("submit-login"));

    expect(screen.getByText("Email Inválido")).toBeTruthy();
  });

  test("Should render the invalid email error and required password error", async () => {
    const { user } = setupComponent(<SignIn />);

    await user.type(screen.getByTestId("email-input"), "invalidEmail.com");
    await user.type(screen.getByTestId("password-input"), "");
    await user.press(screen.getByTestId("submit-login"));

    expect(screen.getByText("Email Inválido")).toBeTruthy();
    expect(screen.getByText("Requerido")).toBeTruthy();
  });

  test("Should call the login function when the form is submitted with valid data", async () => {
    const { user } = setupComponent(<SignIn />);
    const login = jest.spyOn(firebaseAuth, "signInWithEmailAndPassword");

    await user.type(screen.getByTestId("email-input"), "validemail@asd.com");
    await user.type(screen.getByTestId("password-input"), "123456");
    await user.press(screen.getByTestId("submit-login"));

    expect(login).toHaveBeenCalledTimes(1);
  });

  test("Should not call the login function when the form is submitted with invalid data", async () => {
    const { user } = setupComponent(<SignIn />);
    const login = jest.spyOn(firebaseAuth, "signInWithEmailAndPassword");

    await user.type(screen.getByTestId("email-input"), "lorem");
    await user.type(screen.getByTestId("password-input"), "123");
    await user.press(screen.getByTestId("submit-login"));

    expect(login).toHaveBeenCalledTimes(0);
  });

  test("Should show an error message when the login function fails", async () => {
    const { user } = setupComponent(<SignIn />);
    (firebaseAuth.signInWithEmailAndPassword as jest.Mock).mockRejectedValue(
      new Error("invalid-email")
    );

    const toast = jest.spyOn(Toast, "show");
    const handleErrorMessagee = jest.spyOn(localAuth, "useAuthErrorMessage");

    await user.type(screen.getByTestId("email-input"), "validemail@asd.com");
    await user.type(screen.getByTestId("password-input"), "123456");
    await user.press(screen.getByTestId("submit-login"));

    expect(toast).toHaveBeenCalledTimes(1);
    expect(handleErrorMessagee).toHaveBeenCalledTimes(1);
  });
});
