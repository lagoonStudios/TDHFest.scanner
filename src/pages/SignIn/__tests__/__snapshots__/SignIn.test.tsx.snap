// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> Should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#EFFBF4",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "center",
        "gap": 60,
        "height": "90%",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <View>
        <Image
          alt="Ticket scan logo"
          source={1}
        />
      </View>
      <View
        style={
          {
            "display": "flex",
            "gap": 20,
            "width": "70%",
          }
        }
      >
        <View>
          <View
            focusable={true}
            onBlur={[Function]}
            onFocus={[Function]}
            style={
              [
                {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "elevation": 10,
                  "height": "auto",
                  "shadowColor": "#000000",
                  "shadowOffset": {
                    "height": 4,
                    "width": 0,
                  },
                  "shadowOpacity": 0.25,
                  "shadowRadius": 4,
                  "width": "100%",
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#0000004d",
                  "fontWeight": "600",
                  "marginLeft": 15,
                  "marginTop": 10,
                }
              }
            >
              Correo electrónico
            </Text>
            <View>
              <TextInput
                inputMode="email"
                onBlur={[Function]}
                onChangeText={[Function]}
                style={
                  [
                    {
                      "margin": 15,
                    },
                  ]
                }
                testID="email-input"
                value=""
              />
            </View>
          </View>
          <View />
        </View>
        <View>
          <View
            focusable={true}
            onBlur={[Function]}
            onFocus={[Function]}
            style={
              [
                {
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 10,
                  "borderWidth": 2,
                  "elevation": 10,
                  "height": "auto",
                  "shadowColor": "#000000",
                  "shadowOffset": {
                    "height": 4,
                    "width": 0,
                  },
                  "shadowOpacity": 0.25,
                  "shadowRadius": 4,
                  "width": "100%",
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#0000004d",
                  "fontWeight": "600",
                  "marginLeft": 15,
                  "marginTop": 10,
                }
              }
            >
              Password
            </Text>
            <View>
              <TextInput
                onBlur={[Function]}
                onChangeText={[Function]}
                secureTextEntry={true}
                style={
                  [
                    {
                      "margin": 15,
                    },
                  ]
                }
                testID="password-input"
                value=""
              />
            </View>
          </View>
          <View
            style={
              {
                "height": 22,
              }
            }
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#2FF5A3",
              "borderRadius": 10,
              "elevation": 10,
              "opacity": 1,
              "padding": 15,
              "shadowColor": "#000000",
              "shadowOffset": {
                "height": 4,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 4,
            }
          }
          testID="submit-login"
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontWeight": "800",
                "textAlign": "center",
              }
            }
          >
            Ingresar
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignContent": "center",
        "alignItems": "center",
        "display": "flex",
        "marginBottom": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#73C69C",
          "textAlign": "center",
        }
      }
    >
      © Ticketscan todos los derechos reservados 2023
    </Text>
    <Text
      style={
        {
          "color": "#73C69C",
          "textAlign": "center",
        }
      }
    >
      Powered by Lagoon Studio
    </Text>
  </View>
</View>
`;
